src/zzip.cpp

```diff
@@ -0,0 +1,882 @@
+                                     content_end
+                                 );
+        if( compressed_size == 0 ) {
+            // Possibly ran out of file space. Figure out how much is left to insert.
+            // Estimate based on comression ratio of data so far.
+            size_t needed = raw_input_remaining * ( static_cast<double>( bytes_written ) / input_consumed );
+            zip->ensure_capacity_for( content_end + needed + 1024 );
+
+            compressed_size = zip->write_file_at(
+                                  file_relative_path,
+                                  std::string_view{ static_cast<const char *>( file->base() ), file->len() },
+                                  content_end
+                              );
+            if( compressed_size == 0 ) {
+                return nullptr;
+            }
+        }
+        new_entries.emplace_back( compressed_entry{ std::move( file_relative_path ), content_end, compressed_size } );
+        bytes_written += compressed_size;
+        input_consumed += file->len();
+        content_end += compressed_size;
+        raw_input_remaining -= file->len();
+    }
+    zip->update_footer( original_footer, content_end, new_entries, /* shrink_to_fit = */ true );
+    return zip;
+}
+
+bool zzip::extract_to_folder( const std::filesystem::path &path,
+                              const std::filesystem::path &folder,
+                              const std::filesystem::path &dictionary )
+{
+    std::shared_ptr<zzip> zip = zzip::load( path, dictionary );
+
+    for( const JsonMember &entry : zip->footer_.get_object( kEntriesKey ) ) {
+        std::filesystem::path filename = std::filesystem::u8path( entry.name() );
+        size_t len = zip->get_file_size( filename );
+        std::filesystem::path destination_file_path = folder / filename;
+        if( !assure_dir_exist( destination_file_path.parent_path() ) ) {
+            return false;
+        }
+        std::shared_ptr<mmap_file> file = mmap_file::map_writeable_file( destination_file_path );
+        if( !file || !file->resize_file( len ) ) {
+            return false;
+        }
+        if( zip->get_file_to( filename, static_cast<std::byte *>( file->base() ), file->len() ) != len ) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool zzip::compact( double bloat_factor )
+{
+    zzip_footer footer{ footer_ };
+    std::optional<zzip_meta> meta_opt = footer.get_meta();
+    if( !meta_opt.has_value() ) {
+        return false;
+    }
+
+    zzip_meta &meta = *meta_opt;
+    size_t max_allowed_size = meta.total_content_size * bloat_factor + 1024;
+    // Are we even going to try to compact?
+    if( file_->len() < max_allowed_size ) {
+        return false;
+    }
+
+    // Can we save at least one physical page?
+    size_t delta = file_->len() - ( meta.total_content_size + 1024 );
+    if( delta < 4 * 1024 ) {
```

Comment:
consider factoring out `4 * 1024` page size into a constant (since it's used like 4 times in the file).
(it doesn't save any typing of course, just weird seeing raw constant repeated that much)

Relatedly, and more importantly, the `1024` size  for the footer size estimate is very magical, and definitely needs its own named constant. I only happen to know what it means because I read the other function with the comment about it first.